drivers: display: ili9340 refactoring

pavlohamov opened the pull request:
- extend ili9340 driver to use config from dts
 - add posbility to use single driver for ili9xxx displays
 - create generic sheild for the ili934x based boards

vanwinkeljan commented on boards/shields/ili9340_generic/ili9340_adafruit_2_8_tftv2.overlay:
```diff
@@ -0,0 +1,37 @@
+		y-offset = <0>;
+		pixel-format = "RGB888";
+
+	    power-ctrl-1 = [23];
```
Keep indentation the same as above (2 tabs)

vanwinkeljan commented on boards/shields/ili9340_generic/ili9340_seeed_2_8_tftv2.overlay:
```diff
@@ -0,0 +1,42 @@
+		y-offset = <0>;
+		pixel-format = "RGB888";
+
+	    power-ctrl-a = [39 2C 00 34 02];
```
Keep indentation the same as above (2 tabs)

vanwinkeljan commented on drivers/display/display_ili9340.c:
```diff
@@ -1,154 +1,152 @@
-/* The number of bytes taken by a RGB pixel */
-#ifdef CONFIG_ILI9340_RGB565
-#define ILI9340_RGB_SIZE 2U
+#if (DT_INST_0_ILITEK_ILI9XXX_TYPE_ENUM == 0)
```
Please remove this code `DT_INST_0_ILITEK_ILI9XXX_TYPE_ENUM` can never be 0 as it is required device parameter.

pavlohamov replied:
As far as i understood how enums in dts works they are numerated in the same way as in C. 
0 - stand for the firts item in the dts enum (which corresponds to ili9163 for this case).
But this is too implicit. string comparison wont work for #if statement. I will try to investiagate what could be done with this magic number

vanwinkeljan replied:
OK I did not know that the DTS enums returned numbers instead of strings, forget the remark

vanwinkeljan commented on drivers/display/display_ili9340.c:
```diff
@@ -1,154 +1,152 @@
-		return -EPERM;
-	}
+#if (RES_VER > ILI9XXX_Y_MAX) && (RES_VER > ILI9XXX_X_MAX)
+#error "Please check height of \"ilitek,"DEV_TYPE""\" device"
```
Better be consistent in error messages; `"ilitek,"DEV_TYPE"` VS `ilitek,ili9340` bellow

vanwinkeljan reviewed (changes requested):
In general this PR looks good to me except some minor remarks.

Could you also update the display sample yaml file so that CI passes?

jfischer-phytec-iot commented on drivers/display/display_ili9340.c:
```diff
@@ -185,120 +183,307 @@ static int ili9340_read(const struct device *dev, const u16_t x,
+	capabilities->current_orientation = data->orientation;
+}
+
+#define EXEC_CMD(cmd) { \
```
This would be hard to debug, return is opaque to the reader. `_` must not be used and anyway name should be ili93xx_exec_cmd(cmd).

pavlohamov replied:
Done

jfischer-phytec-iot commented on drivers/display/display_ili9340.h:
```diff
@@ -9,67 +10,45 @@
- * @param data Device data structure
- */
-void ili9340_lcd_init(struct ili9340_data *data);
+#define ILI9XXX_CMD_SOFTWARE_RESET 0x01
```
Align constants with tabs

jfischer-phytec-iot commented on dts/bindings/display/ilitek,ili9xxx.yaml:
```diff
@@ -0,0 +1,151 @@
+      required: false
+      description: pixel data format
+      enum:
+        - "RGB565"
```
There is still Kconfig menu for that, either DT or Kconfig

jfischer-no reviewed (changes requested):
This patch has to be spited in self contained logical commits, like:
```
- drivers: display_ili9340: fix style, cleanup
- drivers: display_ili9340: fix style, move init function to the bottom
- drivers: display_ili9340: rename to display_ili93xx
- shields: rename/move adafruit_2_8_tft_touch_v2 shield to ili93xx_generic
- drivers: display_ili93xx: obtain resolution and offsets from DT
- drivers: display_ili93xx: obtain panel settings and parameters from DT
- drivers: display_ili93xx: add support for ABCD controller
- shields: ili93xx_generic: add support for ABCD display
```

pavlohamov commented:
@vanwinkeljan @jfischer-phytec-iot  Please excuse some of obvious thins like typo and indent. I wanted to get some feedback about solution in general. Thank you for your comments.

jfischer-phytec-iot commented:
> @vanwinkeljan @jfischer-phytec-iot Please excuse some of obvious thins like typo and indent.

No worry that is what review is for.

pavlohamov commented on drivers/display/display_ili9340.c:
```diff
@@ -1,154 +1,152 @@
-	k_sleep(K_MSEC(5));
-#endif
+	gpio_pin_set(data->cmd,
+			DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 0);
```
Can't agree with that. From datasheet:
> The D/CX is the signal of data/command
> selection. When D/CX=’1’, D [17:0] bits are display RAM data or command’s parameters. When D/CX=’0’, D
> [17:0] bits are commands.
But i understand your confusion. variable is called "cmd". I will rename int to data or ncmd (now shure which option is better)

jfischer-phytec-iot replied:
No,  `DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 1);` is correct because [The D/CX pin of ILI9340 is active low (transmission command byte).](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/display/ilitek%2Cili9340.yaml). No need for ncmd or cmd.

pavlohamov commented on drivers/display/display_ili9340.c:
```diff
@@ -1,154 +1,152 @@
+		tx_buf.buf = tx_data;
+		tx_buf.len = tx_len;
+		gpio_pin_set(data->cmd,
+			       DT_INST_0_ILITEK_ILI9XXX_CMD_DATA_GPIOS_PIN, 1);
```
> The D/CX is the signal of data/command
> selection. When D/CX=’1’, D [17:0] bits are display RAM data or command’s parameters. When D/CX=’0’, D
> [17:0] bits are commands.
 But variable definitely has to be renamed

pavlohamov committed:
drivers: display_ili9340: fix style, cleanup

 - use IS_ENABLED instead of #ifdef
 - apply new gpio api benefits
 - shorten gpio pin names
 - store pix format and orientation in drv data
 - define magic numbers

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9340: unify names, move init functions

 - ili9340_display -> ili9340_
 - move transmit to the top
 - move init function to the bottom

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9340: define dts instances, add reset func

 - use shorter names for DT_* instances
 - add reset function
 - use shorter branch

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9340: rename to display_ili93xx

change name to use single driver with diferent controllers

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
shields: rename adafruit_2_8_tft_touch_v2 to ili9xxx_generic

target shield is ili9340_adafruit_2_8_tft_touch_v2

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9xxx: obtain display settings from DT

 - pixel format, widht, height and offset are get from DT
 - pixel format enum corresponds to `enum display_pixel_format`

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9xxx: obtain panel settings and parameters from DT

panel config is applied only if value is present

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
shields: move ili9340_adafruit* confit to dts, add ili9340_seeed_tft_v2

move lcd initialization from C code to dts

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov committed:
drivers: display_ili9xxx: cleanup

 - remove API function
 - rename init function
 - hide transmit function (static)

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

pavlohamov commented on drivers/display/display_ili9340.c:
```diff
@@ -185,120 +183,307 @@ static int ili9340_read(const struct device *dev, const u16_t x,
+{
+	struct ili9xxx_data *data = (struct ili9xxx_data *)dev->driver_data;
+
+#ifdef DT_INST_0_ILITEK_ILI9XXX_POWER_CTRL_A
```
Done

pavlohamov commented:
i've spitted single commit to multiple, Fixing comment during this process. Also i've removed addition of new controller type. I want to do it in separate PR.
@vanwinkeljan @jfischer-phytec-iot Could you be so kind to take a look?

pavlohamov commented:
@jfischer-phytec-iot could you be so kind to take a look?

pavlohamov commented:
Rebased to latest master

jfischer-no reviewed (changes requested):
> @jfischer-phytec-iot could you be so kind to take a look?

Sorry for delay, I need few more days for review/test.

jfischer-phytec-iot commented on drivers/display/display_ili9340.c:
```diff
@@ -8,37 +8,31 @@
-	struct device *reset_gpio;
+	struct device *rst;
 #endif
-	struct device *command_data_gpio;
```
I agree that it was to long and you probably changed it to resolve coding style issues. But it would be better to keep reset_gpio + cs_gpio, and change command_data_gpio to cmd_gpio.

You put to many different changes in one commit. Sorry, but it is hard to review.

pavlohamov replied:
Form my point of view it is excessive, vars are used with gpio_* functions indicating what kind of device is it. In general in makes sense, but here we have a particular driver where "cmd" is name of pin (cnd - command not data whoul be even better) .

So sorry, there are a lot of changes, which depend on each other

jfischer-phytec-iot commented on drivers/display/display_ili9340.c:
```diff
@@ -62,46 +65,46 @@ static int ili9340_init(struct device *dev)
-	gpio_pin_configure(data->command_data_gpio, DT_INST_0_ILITEK_ILI9340_CMD_DATA_GPIOS_PIN,
-			   GPIO_OUTPUT |
+	gpio_pin_configure(data->cmd, DT_INST_0_ILITEK_ILI9340_CMD_DATA_GPIOS_PIN,
+			   GPIO_OUTPUT_ACTIVE | GPIO_ACTIVE_LOW |
```
Not necessary to change it, the flag like GPIO_ACTIVE_LOW would be obtained from *_GPIOS_FLAGS

pavlohamov replied:
Make sense. My idea was to "force' inversion, but with this change nobody will able to change. thnx, will remove

jfischer-phytec-iot commented on drivers/display/display_ili9340.c:
```diff
@@ -8,37 +8,31 @@
 #include <drivers/spi.h>
 #include <string.h>
 
+#define RES_X_MAX 320U
```
This change should be in own commit and use DT height and width properties.

pavlohamov replied:
This is not related to DT config. This high bound for the ili9xxx drivers.

jfischer-phytec-iot commented on drivers/display/display_ili9xxx.c:
```diff
@@ -16,18 +16,50 @@ LOG_MODULE_REGISTER(display_ili9xxx, CONFIG_DISPLAY_LOG_LEVEL);
+#define ILI9XXX_PIX_FMT_MASK \
+	(PIXEL_FORMAT_RGB_888 | PIXEL_FORMAT_RGB_565 | PIXEL_FORMAT_BGR_565)
 
+#define DT_CS_PORT        DT_INST_0_ILITEK_ILI9XXX_CS_GPIOS_CONTROLLER
```
It does not belong in this commit, please try to move all of the changes regarding DT and style to one commit or wait for #23245

pavlohamov replied:
Could you be so kind to explain why? This is done inside of "drivers: display_ili9xxx: obtain display settings from DT"

gmarull commented:
@pavlohamov any interest in moving this PR further? I'd be happy to take on this work as I'd like to improve ILI9340 driver and it seems that you already have done multiple of the necessary improvements.

pavlohamov commented:
@gmarull, thank you for taking interest into this PR. Unfortunate I have no time for this, please take all my changes and modify them as you will

jfischer-phytec-iot commented:
Close as it is superfluous after #28410 has been merged.
